### ISSUE TITLE
Map<String, Object> as a @Bean not working as expected even with a @Qualifier when autowired.

### ISSUE THREAD

**Issue by vikuwo** [spring-projects/spring-framework#33537]
**Affects:** 6.1.0

---
<u>Code</u>:
```

import java.util.Map;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.beans.factory.annotation.Qualifier;
import org.springframework.boot.CommandLineRunner;
import org.springframework.boot.SpringApplication;
import org.springframework.boot.autoconfigure.SpringBootApplication;
import org.springframework.context.ApplicationContext;
import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Configuration;

@SpringBootApplication
public class Main implements CommandLineRunner {

  @Configuration
  public static class Config {

    @Bean("qualifiedMap")
    Map<String, Object> qualifiedMap() {
      return Map.of("k1", "v1", "k2", "v2", "k3", "v3");
    }

  }

  @Autowired
  @Qualifier("qualifiedMap")
  private Map<String, Object> qualifiedMap;

  @Autowired
  private ApplicationContext applicationContext;
  
  @Override
  public void run(String... args) {
    System.out.printf("Autowired Bean: %s%n", qualifiedMap);
    System.out.printf("Context Bean: %s%n", applicationContext.getBean("qualifiedMap", Map.class));
  }

  public static void main(String[] args) {
    SpringApplication.run(Main.class, args);
  }

}

```

<u>Output</u>:
```

Autowired Bean: {qualifiedMap={k1=v1, k3=v3, k2=v2}}
Context Bean: {k1=v1, k3=v3, k2=v2}

```

When we mark a Map<String, Object> as @Bean with a qualifier, @Autowired refers to a different map. The bean seems fine when directly fetched from the application context.

**Comment by liua1004** [comment]
It will be fine to use @Resource,  `@Autowired` will find candidate by type.  

Map, Collection(include Set and List) and Array are different when autowiring. They will find bean which has the same type as the element they storage.

**Comment by viswanath-kunapalli-work** [comment]
But this shouldn't be the case when we use @Autowire along with a @Qualifier. It should autowire the exact bean.

**Comment by liua1004** [comment]
It depends on how to understand the annotation `Qualifier`.  I think that it just provides a candidate, but not change the autowire logic. Maybe I am wrong.

**Comment by vikuwo** [comment]
At least the below statements should produce the same result. Otherwise the behaviour is not consistent.

```
    System.out.printf("Autowired Bean: %s%n", qualifiedMap);
    System.out.printf("Context Bean: %s%n", applicationContext.getBean("qualifiedMap", Map.class));
```

**Comment by jhoeller** [comment]
As far as I can tell, this works as of 6.2 due to #28122. Feel free to try the recently released 6.2.0-RC1 to double-check.
